LamestWebserver/LamestWebserver/Caching/ResponseCache.cs

```diff
@@ -0,0 +1,459 @@
+        /// </summary>
+        public double CacheMakeRoom_RemoveByTimePercentage
+        {
+            get
+            {
+                return _cacheMakeRoom_RemoveByTimePercentage;
+            }
+
+            set
+            {
+                _cacheMakeRoom_RemoveByTimePercentage = value.Clamp(0.0, 1.0);
+            }
+        }
+
+        private double _cacheMakeRoom_RemoveByTimePercentage = 0.25;
+
+        /// <summary>
+        /// Makes room when the cache is overflowing.
+        /// </summary>
+        /// <param name="requestedSpace">The requested amount of free space. (Not including CacheMakeRoom_AdditionalFreeSpace)</param>
+        protected virtual void MakeRoom(ulong requestedSpace)
+        {
+            if (!StringResponses.Any())
+                return;
+
+            if (!MaximumStringResponseCacheSize.HasValue)
+                return;
+
+            KeyValuePair<string, ResponseCacheEntry<string>>[] responses;
+
+            using (StringResponseLock.LockRead())
+                responses = StringResponses.ToArray();
+
+            // Remove outdated.
+            foreach (var kvpair in responses)
+                if (kvpair.Value.RefreshTime != null && kvpair.Value.LastUpdatedDateTime + kvpair.Value.RefreshTime < DateTime.UtcNow)
+                    RemoveStringResponseEntry(kvpair.Key, kvpair.Value);
+
+            DateTime upperTenthPercentileDate;
+            ulong upperTenthPercentileCount;
+
+            KeyValuePair<string, ResponseCacheEntry<string>>[] DateSorted = null, CountSorted = null, SizeSorted = null;
+
+            using (StringResponseLock.LockWrite())
+            {
+                var worker0 = ThreadedWorker.CurrentWorker.Instance.EnqueueJob((Action)(() => DateSorted = StringResponses.OrderBy((kvpair) => kvpair.Value.LastRequestedDateTime).ToArray()));
+                var worker1 = ThreadedWorker.CurrentWorker.Instance.EnqueueJob((Action)(() => CountSorted = StringResponses.OrderBy((kvpair) => kvpair.Value.Count).ToArray()));
+                var worker2 = ThreadedWorker.CurrentWorker.Instance.EnqueueJob((Action)(() => SizeSorted = StringResponses.OrderBy((kvpair) => kvpair.Value.Response.Length).ToArray()));
+
+                ThreadedWorker.JoinTasks(worker0, worker1, worker2);
+
+                upperTenthPercentileDate = DateSorted[(int)(DateSorted.Length * (1.0 - CacheMakeRoom_UpperPercentile_Date))].Value.LastRequestedDateTime;
+                upperTenthPercentileCount = CountSorted[(int)(CountSorted.Length * (1.0 - CacheMakeRoom_UpperPercentile_Count))].Value.Count;
+                
+                // Remove by size if not in upper tenth percentile.
+                for (int i = (int)(SizeSorted.Length * CacheMakeRoom_RemoveBySizePercentage); i >= 0; i--)
+                {
+                    if (CurrentStringResponseCacheSize + requestedSpace <= MaximumStringResponseCacheSize.Value) // First run without additional size
+                        return;
```

Comment:
we don't wanna return yet! the additional available space is probably not free yet.